### PR TITLE
Remove StringUtil class and update tests references #4819

### LINKED ISSUE #4819: Remove `StringUtil.Compare()` and update references in the tests
The [`StringUtil.Compare()`](https://github.com/nunit/nunit/blob/main/src/NUnitFramework/framework/Internal/StringUtil.cs#L20) function is public but lives in the "Internal" namespace. It isn't actually used in any constraints, but there are three references in the tests still to it. A comment on the class indicates that 

> Provides methods to support legacy string comparison methods.

I'd like to propose we begin the process of removing this. I think we can convert the test usage already but I'd like to suggest we mark the method itself as obsolete rather than outright removal even though it's in the internal namespace to maximize others' chance to move away from it.

@nunit/framework-team I think this could be a `good first issue` if we can agree on if we should obsolete or remove outright

### PR DESCRIPTION
**Pull request by mr-sergito**
Fixes #4819

Proceeded to removing the whole internal [`StringUtil`](https://github.com/nunit/nunit/blob/bf2d0f25ab9ccf2b2c6b11f3ec01bd8898fc93e6/src/NUnitFramework/framework/Internal/StringUtil.cs#L10) class and replaced its methods references at the tests for built-in `System.StringComparison`.

### CODE REVIEW

**Comment by mr-sergito** [comment]
> @mr-sergito please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @dotnet-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @dotnet-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @dotnet-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> 
> This Contribution License Agreement ( **“Agreement”** ) is agreed to by the party signing below ( **“You”** ), and conveys certain license rights to the .NET Foundation ( **“.NET Foundation”** ) for Your contributions to .NET Foundation open source projects. This Agreement is effective as of the latest signature date below.
> 
> **1. Definitions.**
> 
> **“Code”** means the computer software code, whether in human-readable or machine-executable form, that is delivered by You to .NET Foundation under this Agreement.
> 
> **“Project”** means any of the projects owned or managed by .NET Foundation and offered under a license approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
> 
> **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any Project, including but not limited to communication on electronic mailing lists, source code control systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of discussing and improving that Project, but excluding communication that is conspicuously marked or otherwise designated in writing by You as “Not a Submission.”
> 
> **“Submission”** means the Code and any other copyrightable material Submitted by You, including any associated comments and documentation.
> 
> **2. Your Submission.** You must agree to the terms of this Agreement before making a Submission to any Project. This Agreement covers any and all Submissions that You, now or in the future (except as described in Section 4 below), Submit to any Project.
> 
> **3. Originality of Work.** You represent that each of Your Submissions is entirely Your original work. Should You wish to Submit materials that are not Your original work, You may Submit them separately to the Project if You (a) retain all copyright and license information that was in the materials as you received them, (b) in the description accompanying your Submission, include the phrase "Submission containing materials of a third party:" followed by the names of the third party and any licenses or other restrictions of which You are aware, and (c) follow any other instructions in the Project's written guidelines concerning Submissions.
> 
> **4. Your Employer.** References to “employer” in this Agreement include Your employer or anyone else for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your Submission is made in the course of Your work for an employer or Your employer has intellectual property rights in Your Submission by contract or applicable law, You must secure permission from Your employer to make the Submission before signing this Agreement. In that case, the term “You” in this Agreement will refer to You and the employer collectively. If You change employers in the future and desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement and secure permission from the new employer before Submitting those Submissions.
> 
> **5. Licenses.**
> 
> **a. Copyright License.** You grant .NET Foundation, and those who receive the Submission directly or indirectly from .NET Foundation, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third parties.
> 
> **b. Patent License.** You grant .NET Foundation, and those who receive the Submission directly or indirectly from .NET Foundation, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under Your patent claims that are necessarily infringed by the Submission or the combination of the Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and import or otherwise dispose of the Submission alone or with the Project.
> 
> **c. Other Rights Reserved.** Each party reserves all rights not expressly granted in this Agreement. No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are granted by implication, exhaustion, estoppel or otherwise.
> 
> **6. Representations and Warranties.** You represent that You are legally entitled to grant the above licenses. You represent that each of Your Submissions is entirely Your original work (except as You may have disclosed under Section 3 ). You represent that You have secured permission from Your employer to make the Submission in cases where Your Submission is made in the course of Your work for Your employer or Your employer has intellectual property rights in Your Submission by contract or applicable law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You have the necessary authority to bind the listed employer to the obligations contained in this Agreement. You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES EXPRESSLY STATED IN SECTIONS 3, 4, AND 6 , THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 
> **7. Notice to .NET Foundation.** You agree to notify .NET Foundation in writing of any facts or circumstances of which You later become aware that would make Your representations in this Agreement inaccurate in any respect.
> 
> **8. Information about Submissions.** You agree that contributions to Projects and information about contributions may be maintained indefinitely and disclosed publicly, including Your name and other information that You submit with Your Submission.
> 
> **9. Governing Law/Jurisdiction.** This Agreement is governed by the laws of the State of Washington, and the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County, Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all defenses of lack of personal jurisdiction and forum non-conveniens.
> 
> **10. Entire Agreement/Assignment.** This Agreement is the entire agreement between the parties, and supersedes any and all prior agreements, understandings or communications, written or oral, between the parties relating to the subject matter hereof. This Agreement may be assigned by .NET Foundation.
> 
> _.NET Foundation dedicates this Contribution License Agreement to the public domain according to the Creative Commons CC0 1._

@dotnet-policy-service agree

File: src/NUnitFramework/framework/Constraints/PathConstraint.cs [hunk 1]
```diff
@@ -117,7 +116,10 @@ protected bool IsSubPath(string path1, string path2)
                 return false;
 
             // path 2 is longer than path 1: see if initial parts match
-            if (!StringUtil.StringsEqual(path1, path2.Substring(0, length1), caseInsensitive))
+            StringComparison comparisonType = caseInsensitive
+                ? StringComparison.CurrentCultureIgnoreCase
+                : StringComparison.CurrentCulture;
```
Comment (manfred-brands) [inline src/NUnitFramework/framework/Constraints/PathConstraint.cs:121]: I know this is a literal transpose from the `StringUtil` class, but would wonder if we should use `Ordinal`.
I don't think that file system paths are cultural aware.

**Review by manfred-brands — COMMENTED** [review]
In my opinion _path_ comparison should be using `Ordinal` comparison.
But I would like a second opinion before I request changes. @OsirisTerje ?

**Comment by stevenaw** [comment]
I would typically be cautious about changing the comparison, however after a bit of investigation it would seem that the current recommendations are for these comparisons to be ordinal (https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings).

I did find this mono issue from a few years back where it's alluded that this culture-specific comparison may be a relic of older Windows versions. In the thread it's then suggested that, at the time, mono should not make that change "blindly" on their end and that instead caution should be taken (https://github.com/mono/mono/issues/10230#issuecomment-416812677).

All that is to say I'm also curious @OsirisTerje 's thoughts on the matter

File: src/NUnitFramework/framework/Constraints/PathConstraint.cs [hunk 2]
```diff
@@ -117,7 +116,10 @@ protected bool IsSubPath(string path1, string path2)
                 return false;
 
             // path 2 is longer than path 1: see if initial parts match
-            if (!StringUtil.StringsEqual(path1, path2.Substring(0, length1), caseInsensitive))
+            StringComparison comparisonType = caseInsensitive
```
Comment (stevenaw) [inline src/NUnitFramework/framework/Constraints/PathConstraint.cs:119]: One suggestion: As `SamePathConstraint` and `SamePathOrUnderConstraint` both inherit from `PathConstraint` it may help avoid duplication if we exposed a `protected` property or method here to encapsulate the case-sensitivity check and return a `StringComparison` for use.

**Review by stevenaw — CHANGES_REQUESTED** [review]
Thanks @mr-sergito ! The changes look great. I have left one minor suggestion outside of the Ordinal vs culture discussion to help avoid duplication while still allowing us to remove the `StringUtil` class.

**Comment by OsirisTerje** [comment]
Being a Norwegian we do have locale issues.  We should not use CurrentCulture, as that may change dependent upon locale installation.  Since this is for Paths, Ordinal would be the preferred.  Imho.  

But, since this is a PathConstraint and part of the framework, may it have some effects on someones tests if we change it?  We don't have any issues for the PathConstraint  itself (open nor closed) as far as I can see.

**Comment by manfred-brands** [comment]
> But, since this is a PathConstraint and part of the framework, may it have some effects on someones tests if we change it? We don't have any issues for the PathConstraint itself (open nor closed) as far as I can see.

We currently have a non-consistent system due to changes in culture comparison in [.net 6](https://learn.microsoft.com/en-us/dotnet/core/compatibility/globalization/6.0/culture-creation-invariant-mode)

Using the German culture (de-DE), .NET Framework treats 'ss' as being equal to 'ß'. But .NET 6+ does not, neither does the file system. The below test passes in the first, fails in the latter.

```csharp
[Test]
[SetCulture("de-DE")]
public void PathNameComparison()
{
    Assert.That("Grass", Is.SamePath("Graß"));
}
```

But when creating files with those two names, regardless of Framework, I get two files:
```csharp
[Test]
[SetCulture("de-DE")]
public void PathContentComparison()
{
    const string Contents = "Written";
    const string OtherContents = "But not really";
    File.WriteAllText("Graß", OtherContents);
    File.WriteAllText("Grass", Contents);
    string content = File.ReadAllText("Graß");
    Assert.That(content, Is.EqualTo(Contents));
}
```

To me this means two things:
1. Even though we used CurrentCulture we have no tests for this (special non-ascii characters) as otherwise we would have notice .NET core difference.
2. File System is not Culture aware (but  then my OS is en-AU)

So I think we should use Ordinal(IgnoreCase) file file system comparison.

**Comment by stevenaw** [comment]
I'm alright with converting these to ordinal too. I haven't been able to find any evidence of a present or historical OS filesystem which relies on culture. Based on that, we can likely infer that any _tests_ which rely on culture are rare at best, and any that do would likely not accurately test the underlying system. 

Lastly, I have to wonder if the culture comparisons in this helper we're originally added for other reasons. Some .net functions like `StartsWith(string)` and `EndsWith(string)` are themselves culture-aware and so it's possible the culture check were originally added in the NUnit helper to accurately test the corresponding assertion. I haven't done the code archeology to confirm, but it could explain how culture comparisons also came to be used on path-specific use cases.

In summary, and based on the reactions to @manfred-brands comment above, I think the team agrees we should convert these to ordinal 🙂

File: src/NUnitFramework/framework/Constraints/PathConstraint.cs (hunk 2 above)
Comment (mr-sergito) [reply src/NUnitFramework/framework/Constraints/PathConstraint.cs:119]: Hey @stevenaw ! Thanks for the review, I'll make this improvement right away.

**Comment by mr-sergito** [comment]
Thanks everyone for the insights! I honestly didn't know that much about string comparisons, it has been interesting.

I've made the suggested changes regarding string comparisons on path-related tests.

**Review by stevenaw — APPROVED** [review]
This looks great to me, thanks for your contribution @mr-sergito !

@manfred-brands I'll wait a few days to merge to allow you the chance to review if you wanted given you helped shape direction of this too

**Comment by stevenaw** [comment]
Looks like a test failure on the Mac agent. I'm having a hard time seeing the failure on my phone but I'll try again from a computer later on

**Review by stevenaw — APPROVED** [review]
CI is happy again

**Review by manfred-brands — APPROVED** [review]
I also have no further change requests.
Thanks again @mr-sergito for your cooperation.
